### PR TITLE
Support self containing structs in the nodeset compiler

### PR DESCRIPTION
**Pull request by basyskom-jvoe**
This is my suggestion on how to handle the problem described in #3973.

There are two important things to observe with self containing structs:
- The field must be an array (which means that the member will become a pointer)
- A forward declaration must be added instead of doing definition and typedef in one step

The second commit in this PR is related to problems that occur if a struct has a field named "Double", "Int", "Float" or "Char".

### CODE REVIEW

**Comment by basyskom-meerkoetter** [comment]
@jpfr any feedback on this PR? How can we proceed/get this in?

**Comment by jpfr** [comment]
Looks good. How to get it in:
- Make it a PR for master 
- Add a minimal example to the nodeset compiler unit tests

**Comment by basyskom-meerkoetter** [comment]
There is no way to get this into the 1.1 series?

We are fine in the long-run by getting it into master because over time the project will probably move to 1.2, ... but currently we are sitting on 1.1 (and are maintaining this patch).

**Comment by jpfr** [comment]
1.1 is feature frozen.
1.2 might still be possible if we don’t break the API.

**Comment by jpfr** [comment]
That is, 1.1 saw a few changes over time.
But mostly to unblock certification issues.

**Comment by basyskom-meerkoetter** [comment]
1.2 would be fine

**Comment by basyskom-jvoe** [comment]
I just discovered an additional problem: In addition to the union issues in #4042, arrays in unions don't work because the size value and the data pointer are generated inside the same union.

**Comment by basyskom-meerkoetter** [comment]
@jpfr we would like to get this into 1.2. Can you help us there (with a review)?

**Comment by prater27** [comment]
Hi! First of all thanks to the people working in this PR, it would be great making the unions work in custom information models generated with the nodeset compiler.

In regard to this PR, I tried to use some modifications you did some days ago (around ten days I think?). I tried to publish (and subscribe) a custom DataType which is a structure containing among other things an array of unions. The unions in the array are a union of custom DataTypes (Structures), including arrays, etc.

Internally in the server so to say, it worked, so you could call UA_Server_readValue for a variable of the described curstom DataType containing an array of unions (previous to this PR it didn't work and gave memory errors, but after the modifications introduced by this PR it did).
 Nevertheless, I was not able to read the variable correctly from a client (UA_Expert for example), and I could not succesfully subscribe to the published information. I think the problem might had to do with the serialization.

Does this "new" version of the PR include further modifications with respect the previous state of the PR? I mean if some additional modifications were done with respect to "normal" (non self-containing) unions. Otherwise I think there might be something not working perfect yet in regular unions.

If you want, I could provide the information models I could not make sucesfully work with the previous state of the PR (with the modified PR files of around 10 days ago I think).

Thanks in advance! If I could somehow help just let me know, since I am very interested in this feature (unions generated from custom information model).

**Comment by basyskom-jvoe** [comment]
Many thanks for your feedback.

According to the unit tests for the binary encoding I have added, serialization should work as expected. It would be nice if you could provide your information model so I can check why your custom types don't deserialize as expected with UAExpert.

This PR only deals with the OPC UA Binary encoding at the moment. Maybe this could explain the Publish / Subscribe problems.

**Comment by JorgeBlesaGracia** [comment]
Hi! Thanks for your reply. 

I will test the current version of the PR in the coming days and I will let you know if it worked fine also for me or if I had any problem, and in case I had problems I will send the problematic information model.

**Comment by prater27** [comment]
Hi! Thanks for your reply.

I will test the current version of the PR in the coming days and I will let you know if it worked fine also for me or if I had any problem, and in case I had problems I will send the problematic information model.

**Comment by prater27** [comment]
Hi! So I did a small test. Let me know what are your thoughts about the results I got. I might be doing something wrong, but apparently there is something not right with the DataType.

I used the following information model (desing model). I compiled it with the OPC Fundation Compiler:

```xml
<?xml version="1.0" encoding="utf-8"?>

<ModelDesign
        xmlns:uax="http://opcfoundation.org/UA/2008/02/Types.xsd"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xmlns:ua="http://opcfoundation.org/UA/"
        xmlns:ABCD="https://abcd.org"
        xmlns:xsd="http://www.w3.org/2001/XMLSchema"
        TargetNamespace="https://abcd.org"
        TargetXmlNamespace="https://abcd.org"
        TargetVersion="0.9.0"
        TargetPublicationDate="2019-04-01T00:00:00Z"
        xmlns="http://opcfoundation.org/UA/ModelDesign.xsd">

    <Namespaces>
        <Namespace Name="ABCD" Prefix="ABCD" XmlPrefix="ABCD">https://abcd.org</Namespace>
        <Namespace Name="OpcUa" Version="1.03" PublicationDate="2013-12-02T00:00:00Z" Prefix="Opc.Ua" InternalPrefix="Opc.Ua.Server" XmlNamespace="http://opcfoundation.org/UA/2008/02/Types.xsd" XmlPrefix="OpcUa">http://opcfoundation.org/UA/</Namespace>
    </Namespaces>  


    <!-- ### UnionTest_2Fields_DataType ###-->
     <DataType SymbolicName="ABCD:UnionTest_2Fields_DataType" BaseType="ua:Union" IsAbstract="false" SupportsEvents="true" ValueRank="Scalar">
        <Description>UnionTestDataType</Description>
        <Fields>
            <Field Name="switchField" DataType="ua:UInt32" ModellingRule="Mandatory" ValueRank="Scalar">
                <Description>switchField</Description>
            </Field>
            <Field Name="optionA" DataType="ua:Double" ModellingRule="Mandatory" ValueRank="Scalar">
                <Description>Double</Description>
            </Field>
            <Field Name="optionB" DataType="ua:String" ModellingRule="Mandatory" ValueRank="Scalar">
                <Description>String</Description>
            </Field>
        </Fields>
    </DataType>

</ModelDesign>
```

Then I created the following server (based on the Publish example):
https://gist.github.com/prater27/0a0c0592cbde534f334f08dfa486ce91"

I got the following results when inspecting with UA Expert. I also attach images comparing with how is the same union represented by UA Expert when using the Custom DataType server example (it is supposed to be the same DataType).
- Image 1: Union Test Value - UA Expert+Wireshark Value
- Image 2: Union Test DataType - UA Expert 
- Image 3: Union Example Value - UA Expert+Wireshark 
- Image 4: Union Example DataType - UA Expert 

Union Test - UA Expert+Wireshark Value
Union Test - UA Expert DataType
Union Example - UA Expert+Wireshark Value
Union Example - UA Expert DataType

**Comment by prater27** [comment]
Hi! So I did a small test. I am actually confused, because I am using the modifications in this PR (or I thought so), nevertheless I am getting in the generated source the initial errors with the padding and the count of members. Maybe I am doing something wrong...

I used the following information model (design model). I compiled it with the OPC Fundation Compiler:
```xml
<?xml version="1.0" encoding="utf-8"?>

<ModelDesign
        xmlns:uax="http://opcfoundation.org/UA/2008/02/Types.xsd"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xmlns:ua="http://opcfoundation.org/UA/"
        xmlns:ABCD="https://abcd.org"
        xmlns:xsd="http://www.w3.org/2001/XMLSchema"
        TargetNamespace="https://abcd.org"
        TargetXmlNamespace="https://abcd.org"
        TargetVersion="0.9.0"
        TargetPublicationDate="2019-04-01T00:00:00Z"
        xmlns="http://opcfoundation.org/UA/ModelDesign.xsd">

    <Namespaces>
        <Namespace Name="ABCD" Prefix="ABCD" XmlPrefix="ABCD">https://abcd.org</Namespace>
        <Namespace Name="OpcUa" Version="1.03" PublicationDate="2013-12-02T00:00:00Z" Prefix="Opc.Ua" InternalPrefix="Opc.Ua.Server" XmlNamespace="http://opcfoundation.org/UA/2008/02/Types.xsd" XmlPrefix="OpcUa">http://opcfoundation.org/UA/</Namespace>
    </Namespaces>  


    <!-- ### UnionTest_2Fields_DataType ###-->
     <DataType SymbolicName="ABCD:UnionTest_2Fields_DataType" BaseType="ua:Union" IsAbstract="false" SupportsEvents="true" ValueRank="Scalar">
        <Description>UnionTestDataType</Description>
        <Fields>
            <Field Name="switchField" DataType="ua:UInt32" ModellingRule="Mandatory" ValueRank="Scalar">
                <Description>switchField</Description>
            </Field>
            <Field Name="optionA" DataType="ua:Double" ModellingRule="Mandatory" ValueRank="Scalar">
                <Description>Double</Description>
            </Field>
            <Field Name="optionB" DataType="ua:String" ModellingRule="Mandatory" ValueRank="Scalar">
                <Description>String</Description>
            </Field>
        </Fields>
    </DataType>

</ModelDesign>
```

Then I created the following server (based on the Publish example):
https://gist.github.com/prater27/0a0c0592cbde534f334f08dfa486ce91"

I am getting this generated source. I do annotations compared to the same union provided in the examples of this library (custom data type):

```C++
/* Generated from ABCD.Types.bsd with script /usr/local/share/open62541/tools/generate_datatypes.py
 * on host jbg by user jbg at 2021-01-26 11:54:05 */
#include "types_abcd_opcua_generated.h"

/* UnionTest_2Fields_DataType */
static UA_DataTypeMember UnionTest_2Fields_DataType_members[2] = {
{
    UA_TYPES_DOUBLE, /* .memberTypeIndex */
    offsetof(UA_UnionTest_2Fields_DataType, fields.optionA), /* .padding */ <-------- PADDING SHOULD BE UINT32
    true, /* .namespaceZero */
    false, /* .isArray */
    false  /* .isOptional */
    UA_TYPENAME("OptionA") /* .memberName */
},
{
    UA_TYPES_STRING, /* .memberTypeIndex */
    offsetof(UA_UnionTest_2Fields_DataType, fields.optionB), /* .padding */  <-------- PADDING SHOULD BE UINT32
    true, /* .namespaceZero */
    false, /* .isArray */
    false  /* .isOptional */
    UA_TYPENAME("OptionB") /* .memberName */
},};
const UA_DataType UA_TYPES_ABCD_OPCUA[UA_TYPES_ABCD_OPCUA_COUNT] = {
/* UnionTest_2Fields_DataType */
{
    {1, UA_NODEIDTYPE_NUMERIC, {15002LU}}, /* .typeId */
    {1, UA_NODEIDTYPE_NUMERIC, {15007LU}}, /* .binaryEncodingId */
    sizeof(UA_UnionTest_2Fields_DataType), /* .memSize */
    UA_TYPES_ABCD_OPCUA_UNIONTEST_2FIELDS_DATATYPE, /* .typeIndex */
    UA_DATATYPEKIND_UNION, /* .typeKind */
    false, /* .pointerFree */
    false, /* .overlayable */
    3, /* .membersSize */    <---------------------------------------------------THIS SHOULD BE 2
    UnionTest_2Fields_DataType_members  /* .members */
    UA_TYPENAME("UnionTest_2Fields_DataType") /* .typeName */
},
};

Thanks in advance!

```

**Comment by basyskom-jvoe** [comment]
Hi,

your type definition looks incomplete. There is a switch field and two options, but the options don't reference the switch field at all.
This is the example from the unit tests that seems to be working: https://github.com/open62541/open62541/pull/3979/files#diff-357b63793f5196c2a73938a90486ec5d14ac9fea056d4edf7fc3c6145119e979R60

**Comment by prater27** [comment]
Perfect I will take a look to it now, thank you very much!

**Comment by prater27** [comment]
Hi!
I am not sure, maybe I am doing something wrong, but I think one element (the first one) of the defined unions gets removed. The generated code does not look the same than the code provided in the custom union data type example.

Even in the provided test, the generated code I think is not what it is supposed to be. According to my understanding of the test information model, I would expect a union of a double and an array of the union itself. Nevertheless, the Double dissapears.

Selfcontained union test (bsd file):
```xml
  <opc:StructuredType BaseType="ua:Union" Name="SelfContainingUnion">
    <opc:Field TypeName="opc:UInt32" Name="SwitchField"/>
    <opc:Field SwitchField="SwitchField" TypeName="opc:Double" SwitchValue="1" Name="Double"/>
    <opc:Field SwitchField="SwitchField" TypeName="opc:Int32" SwitchValue="2" Name="NoOfArray"/>
    <opc:Field LengthField="NoOfArray" SwitchField="SwitchField" TypeName="tns:SelfContainingUnion" SwitchValue="2" Name="Array"/>
  </opc:StructuredType>
  ```
 xml NodeSet:
 ```xml
     </UAVariable>
    <UADataType NodeId="ns=1;i=4002" BrowseName="1:SelfContainingUnion">
        <DisplayName>SelfContainingUnion</DisplayName>
        <References>
            <Reference ReferenceType="HasSubtype" IsForward="false">i=12756</Reference>
        </References>
        <Definition Name="1:SelfContainingUnion" IsUnion="true">
            <Field DataType="Double" Name="Double"/>
            <Field DataType="SelfContainingUnion" ValueRank="1" ArrayDimensions="0" Name="Array"/>
        </Definition>
    </UADataType>
 ```
    
The generated code is:

```C++
/**
 * SelfContainingUnion
 * ^^^^^^^^^^^^^^^^^^^
 */
typedef enum {
    UA_SELFCONTAININGUNIONSWITCH_NONE = 0,
    UA_SELFCONTAININGUNIONSWITCH_ARRAY = 1,
    __UA_SELFCONTAININGUNIONSWITCH_FORCE32BIT = 0x7fffffff
} UA_SelfContainingUnionSwitch;
UA_STATIC_ASSERT(sizeof(UA_SelfContainingUnionSwitch) == sizeof(UA_Int32), enum_must_be_32bit);

typedef struct UA_SelfContainingUnion UA_SelfContainingUnion;
struct UA_SelfContainingUnion {
    UA_SelfContainingUnionSwitch switchField;
    union {
        struct {
            size_t arraySize;
            UA_SelfContainingUnion *array;
        } array;
    } fields;
};
```
and 

```C++
/* SelfContainingUnion */
static UA_DataTypeMember SelfContainingUnion_members[1] = {
{
    UA_TYPES_ABCD_OPCUA_SELFCONTAININGUNION, /* .memberTypeIndex */
    offsetof(UA_SelfContainingUnion, fields.array), /* .padding */
    false, /* .namespaceZero */
    true, /* .isArray */
    false  /* .isOptional */
    UA_TYPENAME("Array") /* .memberName */
},};
const UA_DataType UA_TYPES_ABCD_OPCUA[UA_TYPES_ABCD_OPCUA_COUNT] = {
/* SelfContainingUnion */
{
    {2, UA_NODEIDTYPE_NUMERIC, {4002LU}}, /* .typeId */
    {2, UA_NODEIDTYPE_NUMERIC, {0}}, /* .binaryEncodingId */
    sizeof(UA_SelfContainingUnion), /* .memSize */
    UA_TYPES_ABCD_OPCUA_SELFCONTAININGUNION, /* .typeIndex */
    UA_DATATYPEKIND_UNION, /* .typeKind */
    false, /* .pointerFree */
    false, /* .overlayable */
    2, /* .membersSize */
    SelfContainingUnion_members  /* .members */
    UA_TYPENAME("SelfContainingUnion") /* .typeName */
},
};
```
Is it not missing a Double Field in the union, before the struct? Or am I missunderstanding something?

I also tried with normal unions and the same happens.

Am I doing something wrong?

Thanks!!

**Comment by prater27** [comment]
I think these lines remove twice the first element of the union, the first time the SwitchField and the second one the actual first member of the union. 
When removing one of them, the generated code seems to be right. Nevertheless, when I build and run the custom datatypes example with the modifications introduced in this PR, I get an OutOfMemoryError when trying to add the union variable node of the example (the other nodes are added normally).

types_parser.py
```Python
            if self.is_union and child.get("Name") in switch_fields:
                continue
```
backend_open62541_typedefinitions.py
```
        if isUnion:
            members = "static UA_DataTypeMember %s_members[%s] = {" % (idName, len(datatype.members)-1)
......
......
......
            if isUnion and i == 0:
                continue 
```

**Comment by basyskom-jvoe** [comment]
The type description used in the unit test is what was generated by the nodeset compiler at that time. Maybe the last rebase of the branch broke it...

**Comment by JorgeBlesaGracia** [comment]
> I think these lines remove twice the first element of the union, the first time the SwitchField and the second one the actual first member of the union. When removing one of them, the generated code seems to be right.
> 
> Nevertheless, when I build and run the custom datatypes example with the modifications introduced by this PR, I get an OutOfMemoryError when trying to add the union variable node of the example (the other nodes are added normally), but I can add custom unions apparently correctly using the nodeset_compiler. I will do further testing.
> 
> types_parser.py
> 
> ```python
>             if self.is_union and child.get("Name") in switch_fields:
>                 continue
> ```
> 
> backend_open62541_typedefinitions.py
> 
> ```python
>         if isUnion:
>             members = "static UA_DataTypeMember %s_members[%s] = {" % (idName, len(datatype.members)-1)
> ......
> ......
> ......
>             if isUnion and i == 0:
>                 continue 
> ```

By doing this modification (I comment out those lines in types_parser.py so the generated code I get is right), I can read from a Open62541 client correctly a very complex structure containing unions, arrays of unions, etc. 
With PubSub I could not make it work yet, I will do further testing, although in wireshark things look good.
When I read a variable which is union/contains unions with UA Expert, if it is a simple union reads it correctly, but when it is more complex it doesn't, but I do not know if it is Open62541 stack fault or UA Experts fault (since when using the Open62541 client to read the same variable works correctly).

**Comment by prater27** [comment]
> I think these lines remove twice the first element of the union, the first time the SwitchField and the second one the actual first member of the union. When removing one of them, the generated code seems to be right.
> 
> Nevertheless, when I build and run the custom datatypes example with the modifications introduced by this PR, I get an OutOfMemoryError when trying to add the union variable node of the example (the other nodes are added normally), but I can add custom unions apparently correctly using the nodeset_compiler. I will do further testing.
> 
> types_parser.py
> 
> ```python
>             if self.is_union and child.get("Name") in switch_fields:
>                 continue
> ```
> 
> backend_open62541_typedefinitions.py
> 
> ```python
>         if isUnion:
>             members = "static UA_DataTypeMember %s_members[%s] = {" % (idName, len(datatype.members)-1)
> ......
> ......
> ......
>             if isUnion and i == 0:
>                 continue 
> ```

By doing this modification (I comment out those lines in types_parser.py so the generated code I get is right), I can read from a Open62541 client correctly a very complex structure containing unions, arrays of unions, etc.
With PubSub I could not make it work yet, I will do further testing, although in Wireshark things look good.

When I read a variable which is union/contains unions with UA Expert, if it is a simple union reads it correctly, but when it is more complex it doesn't, but I do not know if it is Open62541 stack fault or UA Experts fault (since when using the Open62541 client to read the same variable works correctly).

**Comment by basyskom-jvoe** [comment]
I have pushed an addition to the nodeset compiler unit test. Does this test pass for you if you make the described change in the python code?

**Comment by prater27** [comment]
Hi! I will test it as soon as I can and I will give you feedback. Thanks!

**Comment by basyskom-jvoe** [comment]
The current version of the PR should work out of the box.

**Review by andreasebner — COMMENTED** [review]
Hello @basyskom-jvoe, i looked over your work and it looks very good. I think there is one important missing code adjustment:
You changed the implementation of copyUnion and clearUnion due to the adjusted padding calculation with 'ofsetof'. I agree with this change but we need to adjust our custom_datatype_server example (server_types_custom.c). This file uses a Union which is defined in https://github.com/open62541/open62541/blob/0f583db1ef3ffb1bc1a472f7c3c2a23b8f8af619/examples/custom_datatype/custom_datatype.h#L186-L205. The new copy and clear Union implementation will not work together with this Union definition.

File: tests/nodeset-compiler/testtypes.bsd [hunk 1]
```diff
@@ -57,4 +57,11 @@
     <opc:Field Name="array1" TypeName="opc:Double" LengthField="array1Size" />
   </opc:StructuredType>
 
+  <opc:StructuredType BaseType="ua:Union" Name="SelfContainingUnion">
+    <opc:Field TypeName="opc:UInt32" Name="SwitchField"/>
+    <opc:Field SwitchField="SwitchField" TypeName="opc:Double" SwitchValue="1" Name="Double"/>
+    <opc:Field SwitchField="SwitchField" TypeName="opc:Int32" SwitchValue="2" Name="NoOfArray"/>
+    <opc:Field LengthField="NoOfArray" SwitchField="SwitchField" TypeName="tns:SelfContainingUnion" SwitchValue="2" Name="Array"/>
```
[the remark this answers is not in the document]
Comment (basyskom-jvoe) [reply tests/nodeset-compiler/testtypes.bsd:64]: "NoOfArray" is the LengthField for the "Array" field, so they both need to be enabled together.

**Comment by basyskom-jvoe** [comment]
@andreasebner Many thanks for the review!
Does my last commit fix the issue with the manually added custom type?

**Comment by basyskom-meerkoetter** [comment]
@andreasebner as this PR is important to me, i would like to confirm who is doing what. You have a go from @jpfr to decide to merge it for 1.2 or not. Do you need anything from us?

I don't want to push you, I just don't want to miss the 1.2-boat because we are waiting on each other.

**Comment by andreasebner** [comment]
@basyskom-jvoe Yes your latest fix solves the reported Problem. @basyskom-meerkoetter This PR should definitely be part of our 1.2 release and therefore I will merge it now. Thanks for the contribution!

**Comment by basyskom-meerkoetter** [comment]
Thank you for the review! :-)

**Pull request merged**